LazyCache/AsyncLazy.cs

```diff
@@ -16,7 +16,10 @@ public class AsyncLazy<T> : Lazy<Task<T>>
         }
 
         public AsyncLazy(Func<Task<T>> taskFactory) :
-            base(() => Task.Factory.StartNew(taskFactory).Unwrap())
+            base(() => Task.Factory.StartNew(currentContext => {
+                System.Web.HttpContext.Current = (System.Web.HttpContext) currentContext;

```

Comment:
This doesn't look quite right, since this class can be used in a variety of applications, where the context is not a HttpContext?
Maybe you should create another overload of the constructor instead:
```
        public AsyncLazy(Func<object, Task<T>> taskFactory, object state) :
            base(() => Task.Factory.StartNew(taskFactory, state).Unwrap())
        {
        }
```

Reply:
The alternative you suggest would look like:
```
public AsyncLazy(Func<object, Task<T>> taskFactory, object state) :
            base(() => Task.Factory.StartNew(currentContext => {
                System.Web.HttpContext.Current = (System.Web.HttpContext)currentContext;
                return taskFactory(state);
            }, state).Unwrap())
        {
        }
```
That would be the third contructor. The other two (original) contructors of lazyCache won't be used since only one is called from the CachingService class.
Code in the CachingService.GetOrAddAsync() would be changed to:

```
object CacheFactory(ICacheEntry entry) =>
                new AsyncLazy<T>((httpContext) =>
                {
                    var result = addItemFactory(entry);
                    EnsureEvictionCallbackDoesNotReturnTheAsyncOrLazy<T>(entry.PostEvictionCallbacks);
                    return result;
                }, System.Web.HttpContext.Current);
```

So this alternative passes the HttpContext in the CachingService to a new contructor in AsyncLazy.
It seems a lot more code for the same result.
To answer your question: for applications not having a HttpContext, the value remains null. No errors were thrown.

Reply:
Thanks for more information! 
The use of System.Web does not belong in LazyCache, it is an application matter that should not pollute this library. Next time some other person wants some other context to be saved, or other specific mechanism built in to support his application, etc, which will render the library into a mess.
So we need to solve your problem in another way. Either by a generic extension point or by handling your problem only in your own code instead. 
I assume you need to call GetOrAddAsync from a controller method. Can't you save the context yourself into a local variable in the controller method, and then use that local variable in your addItemFactory delegate?

Reply:
I agree with your arguments about the dependency of this library.
Since the httpContext is lost within LazyCache code, I thought it would be best to provide a fix in the library.
When I debug, code does following:

```
//addItemFactory is a function which needs the httpContext
await AppCache.GetOrAddAsync(key, addItemFactory, absoluteExpiration);
//Go through some extension methods in LazyCache lib
CachingService.GetOrAddAsync<T>()
//...
var result = GetValueFromAsyncLazy<T>(cacheItem, out var valueHasChangedType);
CachingService.GetValueFromAsyncLazy<T>()
//switch case: first case 'AsyncLazy<T>' is matched
//HttpContext still present
return asyncLazy.Value
//Calling asyncLazy.Value results in calling the second constructor of AsyncLazy:
AzyncLazy.cstr(Func<Task<T>> taskFactory)
//This calls:
AzyncLazy.base(() => Task.Factory.StartNew(taskFactory).Unwrap())
//HttpContext is present inside base(), outside StartNew()
//HttpContext is lost inside StartNew()
```
This is conform what other people report:
https://stackoverflow.com/questions/10662456/how-do-i-access-httpcontext-current-in-task-factory-startnew
https://stackoverflow.com/questions/8925227/access-httpcontext-current-from-different-threads

Discussion about the MS implementation of AsyncLazy:
https://github.com/dotnet/runtime/issues/27510

Reply:
Yes, HttpContext is bound to the request thread, but the delegate run by StartNew will in most cases be run in another thread.
That's why I asked if you could do this instead:
```
var httpContext = System.Web.HttpContext.Current;
var addItemFactory = async entry => 
{
    // here you can use httpContext instead of System.Web.HttpContext.Current
};
await AppCache.GetOrAddAsync(key, addItemFactory, absoluteExpiration);
```

Reply:
That does not work in my case. I get the following error when entering the wrapping function:
'The name 'httpContext' does not exist in the current context'.